restore-hooks_product-requirements.md

```diff
@@ -0,0 +1,337 @@
+
+**Note: **Processing of concurrent vs sequential workloads is slated later in the Velero roadmap (see [https://github.com/vmware-tanzu/velero/pull/2548/files](https://github.com/vmware-tanzu/velero/pull/2548/files)). The MVP for this feature set will align with restore of single workloads vs concurrent workload restores. A second epic will be created to address the concurrent restore operations and will be added to the backlog for priority visibility. 
+
+**Note: **Please refer to the Requirements section of this document for more details on what items are P0 (must have in the MVP), P1 (should not ship without for the MVP), P2 (nice to haves).
+
+**<span style="text-decoration:underline;">USE CASE 1</span>**
+
+
+    **Title:** Run restore hook before pod restart.
+
+
+    **Description:  **As a user, I would like to run a restore hook before the applications in my pod are restarted.
+
+**<span style="text-decoration:underline;">______________________________________________________________</span>**
+
+**<span style="text-decoration:underline;">USE CASE 2</span>**
+
+
+    **Title: **Allow restore hook to run on non-kubernetes databases
+
+
+    **Description: **As a user, I would like to run restore hook operations even on databases that are external to kubernetes (such as postgres, elastic, etc…).
+
+**<span style="text-decoration:underline;">______________________________________________________________</span>**
+
+**<span style="text-decoration:underline;">USE CASE 3</span>**
+
+
+    **Title: **Run restore at pod level.
+
+
+    **Description: **As a user, I would like to make sure that I can run the restore hook at the pod level. And, I would like the option to run this with an annotation flag in line or using an init container.**<span style="text-decoration:underline;"> </span>**The restore pre-hook should allow the user to run the command on the container where the pre-hook should be executed. Similar to the backup hooks, this hook should run to default to fun on the first container in the pod. 
+
+**<span style="text-decoration:underline;">______________________________________________________________</span>**
+
+**<span style="text-decoration:underline;">USE CASE 4</span>**
+
+
+    **Title:** Specify container in which to run pre-hook
+
+
+    **Description: **As a user, if I do not want to run the pre-hook command on the first container in the pod (default container), I would like the option to annotate the specific container that i would like the hook to run in.
+
+**<span style="text-decoration:underline;">______________________________________________________________</span>**
+
+**<span style="text-decoration:underline;">USE CASE 5</span>**
+
+
+    **Title: **Check for latest snapshot 
+
+
+    **Description: **As a user, I would like Velero to run a check for the latest snapshot in object storage prior to starting restore operations on a pod. 
```

Comment:
There is a 1:1 mapping between restores and backups and backups have volumesnapshots in them.

On restoring from a backup, the volumesnapshots from that backup are what are used to restore data into the volumes.

IMO, changing that behavior, to choose a volumesnapshot, different from the one in the backup, is out of scope for this feature.

Reply:
This use case is not changing that behavior. Rather, the use case specifies that as part of restore requirements for the product, we want the restore to use the latest volumesnapshot available (latest meaning the volume snapshot that has the most recent timestamp). 

On restoring from backup, we want to make sure that our restore requirements leverage the actions and the 1:1 mapping between restores and backups and that the behavior is not changed.

Reply:
Velero will use the snapshot that is in the backup being restored.